### PR TITLE
[PropertyInfo] Make `PhpDocExtractor::getDocBlock()` public

### PR DESCRIPTION
**Pull request by Nyholm**
| Q             | A
| ------------- | ---
| Branch?       | 7.1
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Issues        | N/a
| License       | MIT

Im parsing properties and I would love to find out what is in the doc block for a property. 
Help me figure out why it would be a bad idea to make this method public.

### CODE REVIEW

**Comment by stof** [comment]
As this method does returns a shape with 3 info and not just the docblock, I would not make it public. I don't think we want to provide BC for the current API.

**Comment by Nyholm** [comment]
Good point. 
I suggest to create a new public method then with a simpler API. What do you think about that?

**Comment by ro0NL** [comment]
> Help me figure out why it would be a bad idea to make this method public.

I think because it's not a component responsibility, as per current contracts.

**Comment by Nyholm** [comment]
Hm.. Im not sure I agree with that. You could expect the "property info" component to be able to extract the Doc Block of a property. 

Or maybe I misunderstand you. Are you suggesting I should introduce a proper interface for this method?

**Comment by ro0NL** [comment]
Shouldn't you be using https://github.com/phpDocumentor/ReflectionDocBlock instead?

**Comment by ro0NL** [comment]
> You could expect the "property info" component to be able to extract the Doc Block of a property

Yes, but there's no real contract for it currently

**Comment by Nyholm** [comment]
> Shouldn't you be using https://github.com/phpDocumentor/ReflectionDocBlock instead?

I looked into that. But I do like the feature from PropertyInfo component that would find where the docblock is. Ie, if the property is not documented it goes to the mutator. 

I'll make sure to prepare a proper contract and prepare this PR.

**Comment by Nyholm** [comment]
I've updated the PR with a proper interface and a good return value.

File: src/Symfony/Component/PropertyInfo/PropertyDocBlockExtractorInterface.php [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+/**
+ * Extract a property's doc block.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+interface PropertyDocBlockExtractorInterface
+{
+    /**
+     * Get doc block for a property.
```
Comment (dunglas) [inline src/Symfony/Component/PropertyInfo/PropertyDocBlockExtractorInterface.php]: I would detail the fallback logic.

Comment (Nyholm) [reply src/Symfony/Component/PropertyInfo/PropertyDocBlockExtractorInterface.php]: Thank you. 

Sure let's put that in the contract too.

**Review by dunglas — APPROVED** [review]

**Review by mtarld — APPROVED** [review]

File: src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php [hunk 2]
```diff
@@ -217,9 +225,11 @@ private function filterDocBlockParams(DocBlock $docBlock, string $allowedParam):
     }
 
     /**
+     * Get DocBlock for a property.
```
Comment (fabpot) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]:
```suggestion
     * Gets a property DocBlock.
```

File: src/Symfony/Component/PropertyInfo/PropertyDocBlockExtractorInterface.php [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+ *
+ * A property's doc block may be located on a constructor promoted argument, on
+ * the property or on a mutator for that property.
+ *
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+interface PropertyDocBlockExtractorInterface
+{
+    /**
+     * Get the first available doc block for a property. It finds the doc block
```
Comment (fabpot) [inline src/Symfony/Component/PropertyInfo/PropertyDocBlockExtractorInterface.php]:
```suggestion
     * Gets the first available doc block for a property. It finds the doc block
```

**Review by fabpot — APPROVED** [review]

**Comment by Nyholm** [comment]
Thank you for the reviews. I've updated according to the suggestions.

File: src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php [hunk 4]
```diff
@@ -217,9 +225,11 @@ private function filterDocBlockParams(DocBlock $docBlock, string $allowedParam):
     }
 
     /**
+     * Gets a property DocBlock.
```
Comment (OskarStark) [inline src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]: It's a little bit confusing, that this and not the `get*` method _gets_ the docblock

You know what I mean? Maybe remove the description entirely?

File: src/Symfony/Component/PropertyInfo/PropertyDocBlockExtractorInterface.php [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+ * @author Tobias Nyholm <tobias.nyholm@gmail.com>
+ */
+interface PropertyDocBlockExtractorInterface
+{
+    /**
+     * Gets the first available doc block for a property. It finds the doc block
+     * by the following priority:
+     * - constructor promoted argument
+     * - the class property
+     * - a mutator method for that property
```
Comment (OskarStark) [inline src/Symfony/Component/PropertyInfo/PropertyDocBlockExtractorInterface.php]:
```suggestion
     *  - constructor promoted argument
     *  - the class property
     *  - a mutator method for that property
```

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpDocExtractorTest.php [hunk 6]
```diff
@@ -43,6 +44,19 @@ public function testExtract($property, ?array $type, $shortDescription, $longDes
         $this->assertSame($longDescription, $this->extractor->getLongDescription('Symfony\Component\PropertyInfo\Tests\Fixtures\Dummy', $property));
     }
 
+    public function testGetDocBlock()
+    {
+        $docBlock = $this->extractor->getDocBlock('Symfony\Component\PropertyInfo\Tests\Fixtures\Dummy', 'g');
```
Comment (OskarStark) [inline src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpDocExtractorTest.php]:
```suggestion
        $docBlock = $this->extractor->getDocBlock(Dummy::class, 'g');
```
Same for the other ones in this file

**Review by OskarStark — APPROVED** [review]
Some comments

File: src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpDocExtractorTest.php (hunk 6 above)
Comment (Nyholm) [reply src/Symfony/Component/PropertyInfo/Tests/Extractor/PhpDocExtractorTest.php]: Thank you!

File: src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php (hunk 4 above)
Comment (Nyholm) [reply src/Symfony/Component/PropertyInfo/Extractor/PhpDocExtractor.php]: Removed

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @Nyholm.

**Pull request merged**